keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -198,49 +197,51 @@ post-filter extension point.
-defined range of integers representing the minimum and maximum scores. After the
-[normalize scoring](#normalize-scoring) phase, the scheduler will combine node
-scores from all plugins according to the configured plugin weights.
+These plugins have two phases:
 
-### Normalize scoring
+1. The first phase is called "score" which is used to rank nodes that have passed 
+the filtering phase. The scheduler will call `Score` of each scoring plugin for
+each node.
+2. The second phase is "normalize scoring" which is used to modify scores before
+the scheduler computes a final ranking of Nodes, and each score plugin receives
+scores given by the same plugin to all nodes in "normalize scoring" phase. 
+`NormalizeScore` is called once per plugin per scheduling cycle right after 
+"score" phase.
 
-These plugins are used to modify scores before the scheduler computes a final
-ranking of Nodes. A plugin that registers for this extension point will be
-called with the [scoring](#scoring) results from the same plugin. This is called
-once per plugin per scheduling cycle.
+The output of a score plugin must be an integer in range of
+**[NodeScoreMin, NodeScoreMax]**. if not, the scheduling cycle is aborted.
+This is the output after running the optional NormalizeScore function of the
+plugin. If NormalizeScore is not provided, the output of Score must be in this range
 
 For example, suppose a plugin `BlinkingLightScorer` ranks Nodes based on how
 many blinking lights they have.
 
 ```go
-func ScoreNode(_ *v1.Pod, n *v1.Node) (int, error) {
-   return getBlinkingLightCount(n)
+func (*BlinkingLightScorer) Score(pc *PluginContext, _ *v1.Pod, nodeName string) (int, *Status) {
+   return getBlinkingLightCount(nodeName)
 }
 ```
 
 However, the maximum count of blinking lights may be small compared to
-`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also register for this
-extension point.
+`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also implement `NormalizeScore`.
 
 ```go
-func NormalizeScores(scores map[string]int) {
+func (*BlinkingLightScorer) NormalizeScore(pc *PluginContext, _ *v1.Pod, nodeScores NodeScoreList) *Status {
    highest := 0
-   for _, score := range scores {
-      highest = max(highest, score)
+   for _, nodeScore := range nodeScores {
+      highest = max(highest, nodeScore.Score)
    }
-   for node, score := range scores {
-      scores[node] = score*NodeScoreMax/highest
+   for i, nodeScore := range nodeScores {
+      nodeScores[i].Score = nodeScore.Score*NodeScoreMax/highest
    }
+   return nil
 }
 ```
 
-If any normalize-scoring plugin returns an error, the scheduling cycle is
-aborted.
+If either `Score` or `NormalizeScore` returns an error, the scheduling cycle is aborted.
 
-**Note:** Plugins wishing to perform "pre-reserve" work should use the
-normalize-scoring extension point.
+**Note:** Plugins wishing to perform "pre-reserve" work should use the `NormalizeScore` in the
+scoring extension point.
```

Comment:
hmm, we don't have a separate NormalizeScore extension point now, so this should be "use the `Score` extension point".

Generally I don't like this recommendation, I would prefer to add a new extension point for "pre-reserve".

Reply:
> hmm, we don't have a separate NormalizeScore extension point now, so this should be "use the `Score` extension point".

It is "use the `NormalizeScore` in the score extension point".

Reply:
What I find awkward is that we are asking them to provide a Score plugin (even if they don't need one) to do pre-reserve work. Now that we combined NormalizeScore with Score, they will have to provide a no-op score function if all they needed to do is a pre-reserve. I suggest we remove this recommendation and add a pre-reserve extension point if we find it necessary in the future.

Reply:
this was resolved without addressing the comment. I think we should remove this note.

Reply:
done

Reply:
Please take another look